### ISSUE TITLE
Is it possible train SSD or YOLO  in fp16?

### ISSUE THREAD

**Issue by LightToYang** [dmlc/gluon-cv#284]
I find that in train_imagenet.py --dtype could be set float16 for faster training,
`parser.add_argument('--dtype', type=str, default='float32',
                    help='data type for training. default is float32')`
but I don't the similar setting in train_ssd.py or train_yolov3.py. So, is it possible train the objection detection module in fp16 now?

**Comment by zhreshold** [comment]
It's doable, but we need to make sure everything goes alright for detection with fp 16.

**Comment by tlind** [comment]
Did anybody succeed with this?

**Comment by zhreshold** [comment]
@stereomatchingkiss This is currently hardcoded as float32: https://github.com/dmlc/gluon-cv/blob/master/gluoncv/loss.py#L222

See if changing this helps.

**Comment by enry12** [comment]
Hello all,

I'm trying to run YOLOv3 using float16 instead of float32 at inference time, I have changed both network and data type as described [here](https://mxnet.incubator.apache.org/versions/master/faq/float16.html). 

Unfortunately, the script throws me an error (pasted below)! In my opinion there is some hardcoded "float32" data type in their implementation. Moreover, I have tried to change some of these hardcoded types without success.

Can someone help me with this problem?

Thank you in advance

----------------------------------------------------------------------------------

...
ids, scores, bboxes = self.yolo(x)
  File "/home/civitelli/applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/gluon/block.py", line 548, in __call__
    out = self.forward(*args)
  File "/home/civitelli/applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/gluon/block.py", line 915, in forward
    return self._call_cached_op(x, *args)
  File "/home/civitelli/applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/gluon/block.py", line 821, in _call_cached_op
    out = self._cached_op(*cargs)
  File "/home/civitelli/applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/_ctypes/ndarray.py", line 150, in __call__
    ctypes.byref(out_stypes)))
  File "/home/civitelli/applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/base.py", line 253, in check_call
    raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: Error in operator yolov30_yolooutputv30_broadcast_add1: [15:42:32] src/operator/contrib/../elemwise_op_common.h:135: Check failed: assign(&dattr, vec.at(i)): Incompatible attr in node yolov30_yolooutputv30_broadcast_add1 at 1-th input: expected float16, got float32

**Comment by enry12** [comment]
Hello all,

I'm trying to run YOLOv3 using float16 instead of float32 at inference time, I have changed both network and data type as described here.

Unfortunately, the script throws me an error (pasted below)! In my opinion there is some hardcoded "float32" data type in their implementation. Moreover, I have tried to change some of these hardcoded types without success.

Can someone help me with this problem?

Thank you in advance

...
ids, scores, bboxes = self.yolo(x)
File ".../applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/gluon/block.py", line 548, in call
out = self.forward(*args)
File ".../applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/gluon/block.py", line 915, in forward
return self._call_cached_op(x, *args)
File ".../applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/gluon/block.py", line 821, in _call_cached_op
out = self._cached_op(*cargs)
File ".../applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/_ctypes/ndarray.py", line 150, in call
ctypes.byref(out_stypes)))
File ".../applications/anaconda3/envs/mxnet/lib/python3.7/site-packages/mxnet/base.py", line 253, in check_call
raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: Error in operator yolov30_yolooutputv30_broadcast_add1: [15:42:32] src/operator/contrib/../elemwise_op_common.h:135: Check failed: assign(&dattr, vec.at(i)): Incompatible attr in node yolov30_yolooutputv30_broadcast_add1 at 1-th input: expected float16, got float32

**Comment by enry12** [comment]
Hi @zhreshold,

It gives the same exception also changing that value to "float16". 
It seems that somewhere this implementation of YOLO cast something to float32 by default.

Have you ever try to run it in float16?

**Comment by enry12** [comment]
Hello all,

I have probably found a way to run YOLOv3 in float16. 
Basically, the problems are some hidden hardcoded float32 tensors, I will put my "corrections" here asap.

**Comment by Jerryzcn** [comment]
@enry12 with mxnet 1.5 you can try amp. it automatically uses fp16, although slightly slower than manual fp16. You can check RCNN's script for amp training. For RCNN, we are able to achieved around 30% speedup

**Comment by Jerryzcn** [comment]
https://github.com/dmlc/gluon-cv/blob/master/scripts/detection/faster_rcnn/train_faster_rcnn.py#L463

**Comment by djaym7** [comment]
+1 @Jerryzcn you can double the batch size 2x fp32 and increase the learning rate as well when you use amp making training much faster!

**Comment by never-to-never** [comment]
Does anyone have a successful modification plan for training YOLOV3 in FP16？

**Comment by enry12** [comment]
Yes, I have successfully trained YOLOv3 with a reduced precision (float 16) some time ago. 
The problem is in the forward function: if I'm not wrong there are some operations that return tensors in fp32. Unfortunately, i cannot find the code I have used to solve the problem but, in any case, you should be able to solve this issue on your own removing the network hybridisation and see where the forward function fails.
